projects/lights_control/test/test_lights_blinker.c

```diff
@@ -0,0 +1,161 @@
+#include <stdlib.h>
+
+#include "interrupt.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#include "lights_blinker.h"
+
+typedef enum {
+  LIGHTS_TEST_EVENT_1 = 0,
+  LIGHTS_TEST_EVENT_2,
+  NUM_LIGHTS_TEST_EVENTS
+} LightsTestEvent;
+
+#define DURATION_SHORT 300
+#define DURATION_LONG 500
+
+void setup_test(void) {
+  interrupt_init();
+  soft_timer_init();
+  event_queue_init();
+}
+
+void teardown_test(void) {}
+
+void test_lights_blinker_init(void) {
+  LightsBlinker blinker;
+  lights_blinker_init(&blinker);
+  TEST_ASSERT_EQUAL(blinker.timer_id, SOFT_TIMER_INVALID_TIMER);
+}
+
+void test_lights_blinker_on(void) {
+  // Testing two blinkers with different times to make sure one happens after the other
+  LightsBlinker blinker_1;
+  LightsBlinker blinker_2;
+
+  TEST_ASSERT_OK(lights_blinker_init(&blinker_1));
+  TEST_ASSERT_OK(lights_blinker_init(&blinker_2));
+
+  TEST_ASSERT_OK(lights_blinker_on(&blinker_1, DURATION_SHORT, LIGHTS_TEST_EVENT_1));
+  TEST_ASSERT_EQUAL(blinker_1.duration, DURATION_SHORT);
+  TEST_ASSERT_EQUAL(LIGHTS_BLINKER_STATE_ON, blinker_1.state);
+  TEST_ASSERT_NOT_EQUAL(blinker_1.timer_id, SOFT_TIMER_INVALID_TIMER);
+
+  Event e;
+  // Wait for the blinker_1 initial event to fire. This happens almost instantly as the event gets
+  // raised upon call to lights_blinker_on.
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_TEST_EVENT_1);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_ON);
+
+  TEST_ASSERT_OK(lights_blinker_on(&blinker_2, DURATION_LONG, LIGHTS_TEST_EVENT_2));
+  TEST_ASSERT_EQUAL(blinker_2.duration, DURATION_LONG);
+  TEST_ASSERT_EQUAL(LIGHTS_BLINKER_STATE_ON, blinker_2.state);
+  TEST_ASSERT_NOT_EQUAL(blinker_2.timer_id, SOFT_TIMER_INVALID_TIMER);
+
+  // Wait for the blinker_2 initial event to fire.
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_TEST_EVENT_2);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_ON);
+
+  // Now we make sure blinker 1 timer goes off before blinker 2
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_TEST_EVENT_1);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_OFF);
+
+  // Wait for the second blinker
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  TEST_ASSERT_EQUAL(e.id, LIGHTS_TEST_EVENT_2);
+  TEST_ASSERT_EQUAL(e.data, LIGHTS_BLINKER_STATE_OFF);
+}
+
+void test_lights_blinker_on_uninitialized(void) {
+  // We shoudln't be able to call blinker_on on an uninitialized blinker, as uninitialized blinkers
+  // might have timer_id's that may be valid timer id's and may be in use by other modules in the
+  // application.
+  LightsBlinker blinker;
+  TEST_ASSERT_NOT_OK(lights_blinker_on(&blinker, DURATION_SHORT, LIGHTS_TEST_EVENT_1));
```

Comment:
This might not actually pass. It's possible that whatever occupied that memory before happened to have a seemingly valid `blinker`. I don't think it's something you need to test for.

Reply:
In order for blinker_on to be called on a blinker, it has to have `SOFT_TIMER_INVALID_TIMER` as its timer id. If it has a valid `timer_id` then this should fail, because that means the blinker is in use.

Reply:
Yes, but uninitialized memory can be set to anything. What you've stated is unrelated to my comment.

Reply:
Oh ya this could fail based on what's in the uninitialized memory. I'll remove this test.